x/shield/keeper/purchase.go

```diff
@@ -137,94 +148,133 @@ func (k Keeper) IterateAllPurchases(ctx sdk.Context, callback func(purchase type
 func (k Keeper) RemoveExpiredPurchases(ctx sdk.Context) {
 	store := ctx.KVStore(k.storeKey)
 	iterator := k.PurchaseQueueIterator(ctx, ctx.BlockTime())
-
+	bondDenom := k.sk.BondDenom(ctx)
 	defer iterator.Close()
 	for ; iterator.Valid(); iterator.Next() {
-		var timeslice []types.PurchaseTxHash
+		var timeslice []types.PPPair
 		k.cdc.MustUnmarshalBinaryLengthPrefixed(iterator.Value(), &timeslice)
-		for _, entry := range timeslice {
-			purchase, _ := k.GetPurchase(ctx, entry)
-			pool, err := k.GetPool(ctx, purchase.PoolID)
-			if err == nil {
-				pool.Available = pool.Available.Add(purchase.Shield.AmountOf(k.sk.BondDenom(ctx)))
-				pool.Shield = pool.Shield.Sub(purchase.Shield)
-				k.SetPool(ctx, pool)
+		for _, ppPair := range timeslice {
+			purchaseList, _ := k.GetPurchaseList(ctx, ppPair.PoolID, ppPair.Purchaser)
+
+			for i := 0; i < len(purchaseList.Entries); i++ {
```

Comment:
Somehow I feel the i++ should be moved into a conditional statement inside and the i-- should be removed, to make the logic more natural.

Reply:
So I guess i++ if we don't splice the array. I guess it's easier that way